### PR TITLE
TT-2437 add mnemonic with-index to kms signatures for LTC and BTC

### PR DESCRIPTION
**Pull request by pedrodbaptista**
- add mnemonic with-index to kms signatures for LTC and BTC
- remove LUNA

### CODE REVIEW

File: src/management.ts [hunk 1]
```diff
@@ -150,6 +148,35 @@ export const generateManagedPrivateKeyBatch = async (
   }
 }
 
+export const getWalletWithMnemonicForChain = async (chain: Currency, path?: string, pwd?: string, print = true) => {
+  const password = pwd ?? config.getValue(ConfigOption.KMS_PASSWORD)
+  const pathToWallet = path || homedir() + '/.tatumrc/wallet.dat'
+  if (!existsSync(pathToWallet)) {
+    console.error(JSON.stringify({ error: `No such wallet for chain '${chain}'.` }, null, 2))
```
Comment (alexloiko) [inline src/management.ts:155]: Let's extract common code of wallet read

File: src/management.ts [hunk 2]
```diff
@@ -150,6 +148,35 @@ export const generateManagedPrivateKeyBatch = async (
...
+    return
+  }
+  const data = readFileSync(pathToWallet, { encoding: 'utf8' })
+  if (!data?.length) {
+    console.error(JSON.stringify({ error: `No such wallet for chain '${chain}'.` }, null, 2))
```
Comment (alexloiko) [inline src/management.ts:160]: please extract common log method, to have just log('some message')

File: src/signatures.ts [hunk 3]
```diff
@@ -335,13 +330,31 @@ const processTransaction = async (
       break
     }
     case Currency.LTC: {
+      const signatures = transaction.signatures;
+      const privateKeys: string[] = wallets.reduce((pkAgg, w) => {
+        signatures?.forEach(async s => {
+          if (w.mnemonic !== undefined && s.index !== undefined && s.index !== null) {
```
Comment (alexloiko) [inline src/signatures.ts:336]: please use _.isNil() from lodash and extract condition to named variable

File: src/management.ts [hunk 4]
```diff
@@ -150,6 +148,35 @@ export const generateManagedPrivateKeyBatch = async (
...
+    return
+  }
+  try {
+    const walletData: Partial<{[key: string]: {chain?: string; mnemonic?: string}}> = JSON.parse(AES.decrypt(data, password).toString(enc.Utf8))
+    const wallets: any[] = Object.values(walletData ?? {}).filter(w => w?.chain === chain && w?.mnemonic !== null && w?.mnemonic !== undefined)
```
Comment (alexloiko) [inline src/management.ts:165]: please use _.isNil()

File: src/signatures.ts [hunk 5]
```diff
@@ -82,17 +82,20 @@ const processTransaction = async (
     }
   }
 
-  const wallets = []
+  let wallets: any[] = []
```
Comment (alexloiko) [inline src/signatures.ts:85]: try not to use explicit any

File: src/signatures.ts [hunk 6]
```diff
@@ -82,17 +82,20 @@ const processTransaction = async (
...
+    wallets.push(...await getWalletWithMnemonicForChain(transaction.chain, path, pwd, false) ?? [])
+  }
   let txData = ''
   console.log(`${new Date().toISOString()} - Processing pending transaction - ${JSON.stringify(transaction, null, 2)}.`)
   switch (transaction.chain) {
     case Currency.ALGO: {
       const algoSecret = wallets[0].secret ? wallets[0].secret : wallets[0].privateKey
       await algorandBroadcast(
-        (await signAlgoKMSTransaction(transaction, algoSecret, testnet))?.txId as string,
+        (await signAlgoKMSTransaction(transaction, algoSecret, testnet)) as string,
```
Comment (alexloiko) [inline src/signatures.ts:98]: is it expected change?

File: src/management.ts (hunk 1 above)
Comment (pedrodbaptista) [reply src/management.ts:155]: makes total sense

File: src/management.ts (hunk 2 above)
Comment (pedrodbaptista) [reply src/management.ts:160]: I already change above, but we can talk about this to have a better logger

File: src/signatures.ts (hunk 3 above)
Comment (pedrodbaptista) [reply src/signatures.ts:336]: done

File: src/management.ts (hunk 4 above)
Comment (pedrodbaptista) [reply src/management.ts:165]: done

File: src/signatures.ts (hunk 5 above)
Comment (pedrodbaptista) [reply src/signatures.ts:85]: just remove it.

File: src/signatures.ts [hunk 7]
```diff
@@ -58,13 +58,34 @@ import {
...
+import _ from 'lodash'
+import { ParcialWallet, Signature } from './interfaces'
+
+function getPrivateKeys(wallets: ParcialWallet[], signatures: Signature[], currency: Currency): string[] {
+  if (!wallets || (wallets && wallets.length === 0)) {
+    return []
+  }
+  return wallets.reduce((pkAgg: string[], w: ParcialWallet) => {
+    if (signatures.length > 0) {
+      signatures?.forEach(async (s: Signature) => {
```
Comment (samuelsramko) [inline src/signatures.ts:73]: dont use .forEach and async -> replace with for...of and await inside.

File: src/signatures.ts [hunk 8]
```diff
@@ -58,13 +58,34 @@ import {
...
 import { broadcast as solanaBroadcast, signKMSTransaction as signSolanaKMSTransaction } from '@tatumio/tatum-solana'
-import { TatumTerraSDK } from '@tatumio/terra'
 import { AxiosInstance } from 'axios'
-import { getManagedWallets, getWallet } from './management'
+import { getManagedWallets, getWallet, getWalletWithMnemonicForChain } from './management'
 import { KMS_CONSTANTS } from './constants'
+import _ from 'lodash'
+import { ParcialWallet, Signature } from './interfaces'
+
+function getPrivateKeys(wallets: ParcialWallet[], signatures: Signature[], currency: Currency): string[] {
```
Comment (samuelsramko) [inline src/signatures.ts:67]: replace with arrow function and const

File: src/signatures.ts [hunk 9]
```diff
@@ -74,317 +95,328 @@ const processTransaction = async (
...
       return
     }
   }
-
   const wallets = []
-  for (const hash of transaction.hashes) {
+  for (const hash of blockchainSignature.hashes) {
     wallets.push(await getWallet(hash, path, pwd, false))
   }
+  if (blockchainSignature.signatures && blockchainSignature.signatures.length > 0) {
```
Comment (samuelsramko) [inline src/signatures.ts:111]: replace with  (blockchainSignature?.signatures?.length > 0)

File: src/signatures.ts [hunk 10]
```diff
@@ -74,317 +95,328 @@ const processTransaction = async (
...
+        await btcBroadcast(await signBitcoinKMSTransaction(blockchainSignature, privateKeys), blockchainSignature.id)
       }
+
       break
     }
     case Currency.LTC: {
-      if (transaction.withdrawalId) {
-        txData = await signLitecoinOffchainKMSTransaction(transaction, wallets[0].mnemonic, testnet)
+      const privateKeys = getPrivateKeys(wallets, signatures, Currency.LTC)
+      console.log('privatekeys', privateKeys, wallets, signatures)
```
Comment (samuelsramko) [inline src/signatures.ts:373]: remove console.log

**Review by samuelsramko — CHANGES_REQUESTED** [review]

File: src/interfaces.ts [hunk 11]
```diff
@@ -0,0 +1,30 @@
+import { Currency } from '@tatumio/tatum'
+
+export interface Signature {
+  id: string
+  index: number
+}
+
+export interface ParcialWallet {
```
Comment (samuelsramko) [inline src/interfaces.ts:8]: dont like naming here - not descriptive enough

File: src/signatures.ts [hunk 12]
```diff
@@ -58,13 +58,39 @@ import {
...
+import _ from 'lodash'
+import { ParcialWallet, Signature } from './interfaces'
+
+const getPrivateKeys = async (
+  wallets: ParcialWallet[],
+  signatures: Signature[],
+  currency: Currency,
+): Promise<string[]> => {
+  const keys: string[] = []
+  if (!wallets || (wallets && wallets.length === 0)) {
```
Comment (samuelsramko) [inline src/signatures.ts:73]: again - if (!wallets?.length)

Comment (pedrodbaptista) [reply src/signatures.ts:73]: make sense, thanks

File: src/interfaces.ts (hunk 11 above)
Comment (pedrodbaptista) [reply src/interfaces.ts:8]: actually you made me noticed that we have same naming in 2 interfaces (need to change both)

**Review by samuelsramko — APPROVED** [review]

**Pull request merged**